File: backend/app/api/v1/endpoints/samples.py
"""Amostras, testes atribuídos e ações do workflow."""

from datetime import datetime
from typing import Annotated

from fastapi import APIRouter, Query, status

from app.api.deps import DbSession
from app.api.v1.endpoints._common import Paging, errors, requires
from app.domain.enums import SamplePriority, SampleStatus
from app.domain.permissions import Permission
from app.schemas.common import Page
from app.schemas.samples import (
    AssignTestsRequest,
    ReasonRequest,
    SampleCreate,
    SampleDetail,
    SampleFilter,
    SampleSummary,
    SampleUpdate,
    StatusHistoryRead,
    to_detail,
    to_history,
    to_summary,
)
from app.services.sample_service import SampleService

router = APIRouter(tags=["Samples"])

Reader = requires(Permission.SAMPLE_READ)
Creator = requires(Permission.SAMPLE_CREATE)
Assigner = requires(Permission.SAMPLE_ASSIGN_TESTS)
Analyst = requires(Permission.SAMPLE_ANALYZE)
Canceller = requires(Permission.SAMPLE_CANCEL)


@router.get("/samples", response_model=Page[SampleSummary], responses=errors(401, 403, 422))
def search_samples(
    session: DbSession,
    _: Reader,
    paging: Paging,
    code: Annotated[str | None, Query(max_length=20, examples=["SMP-2026"])] = None,
    product_id: int | None = None,
    client_id: int | None = None,
    lot_number: Annotated[str | None, Query(max_length=50)] = None,
    status_: Annotated[list[SampleStatus] | None, Query(alias="status")] = None,
    priority: SamplePriority | None = None,
    responsible_id: Annotated[int | None, Query(description="Analista responsável")] = None,
    received_from: datetime | None = None,
    received_to: datetime | None = None,
    q: Annotated[
        str | None, Query(max_length=100, description="Código, lote, produto ou cliente")
    ] = None,
) -> Page[SampleSummary]:
    """Pesquisa paginada. `status` aceita vários valores (`?status=RECEIVED&status=IN_ANALYSIS`)."""
    filters = SampleFilter(
        code=code,
        product_id=product_id,
        client_id=client_id,
        lot_number=lot_number,
        statuses=list(status_ or []),
        priority=priority,
        responsible_id=responsible_id,
        received_from=received_from,
        received_to=received_to,
        q=q,
    )
    result = SampleService(session).search(filters, paging)
    return Page.build([to_summary(s) for s in result.items], result.total, result.page, result.size)


@router.post(
    "/samples",
    response_model=SampleDetail,
    status_code=status.HTTP_201_CREATED,
    responses=errors(401, 403, 422),
    summary="Registra amostra",
    description="Gera o código `SMP-AAAA-NNNN` e atribui os testes do plano analítico do produto.",
)
def create_sample(data: SampleCreate, session: DbSession, actor: Creator) -> SampleDetail:
    return to_detail(SampleService(session).create(data, actor))


@router.get("/samples/{sample_id}", response_model=SampleDetail, responses=errors(401, 403, 404))
def get_sample(sample_id: int, session: DbSession, _: Reader) -> SampleDetail:
    return to_detail(SampleService(session).get(sample_id))


@router.patch(
    "/samples/{sample_id}",
    response_model=SampleDetail,
    responses=errors(401, 403, 404, 409, 422),
    summary="Altera dados de registro",
    description="Exige a `version` lida; se outra pessoa alterou antes, retorna 409.",
)
def update_sample(
    sample_id: int, data: SampleUpdate, session: DbSession, actor: Creator
) -> SampleDetail:
    return to_detail(SampleService(session).update(sample_id, data, actor))


@router.post(
    "/samples/{sample_id}/tests",
    response_model=SampleDetail,
    responses=errors(401, 403, 404, 409, 422),
    summary="Atribui testes adicionais",
)
def assign_tests(
    sample_id: int, data: AssignTestsRequest, session: DbSession, actor: Assigner
) -> SampleDetail:
    return to_detail(
        SampleService(session).assign_tests(sample_id, data.test_definition_ids, actor)
    )


@router.post(
    "/sample-tests/{sample_test_id}/cancel",
    response_model=SampleDetail,
    responses=errors(401, 403, 404, 409, 422),
    summary="Cancela teste pendente (exige justificativa)",
)
def cancel_test(
    sample_test_id: int, data: ReasonRequest, session: DbSession, actor: Assigner
) -> SampleDetail:
    return to_detail(SampleService(session).cancel_test(sample_test_id, data.reason, actor))


@router.post(
    "/samples/{sample_id}/start-analysis",
    response_model=SampleDetail,
    responses=errors(401, 403, 404, 409),
    summary="RECEIVED → IN_ANALYSIS",
)
def start_analysis(sample_id: int, session: DbSession, actor: Analyst) -> SampleDetail:
    return to_detail(SampleService(session).start_analysis(sample_id, actor))


@router.post(
    "/samples/{sample_id}/submit-for-review",
    response_model=SampleDetail,
    responses=errors(401, 403, 404, 409),
    summary="IN_ANALYSIS → AWAITING_REVIEW (todos os testes concluídos)",
)
def submit_for_review(sample_id: int, session: DbSession, actor: Analyst) -> SampleDetail:
    return to_detail(SampleService(session).submit_for_review(sample_id, actor))


@router.post(
    "/samples/{sample_id}/cancel",
    response_model=SampleDetail,
    responses=errors(401, 403, 404, 409, 422),
    summary="Cancela a amostra (exige justificativa)",
)
def cancel_sample(
    sample_id: int, data: ReasonRequest, session: DbSession, actor: Canceller
) -> SampleDetail:
    return to_detail(SampleService(session).cancel(sample_id, data.reason, actor))


@router.get(
    "/samples/{sample_id}/status-history",
    response_model=list[StatusHistoryRead],
    responses=errors(401, 403, 404),
)
def status_history(sample_id: int, session: DbSession, _: Reader) -> list[StatusHistoryRead]:
    return [to_history(entry) for entry in SampleService(session).status_history(sample_id)]

File: frontend/src/api/samples.ts
import type {
  Page,
  ResultEntry,
  ResultListItem,
  ResultRead,
  SampleAction,
  SampleCreate,
  SampleDetail,
  SamplePriority,
  SampleStatus,
  SampleSummary,
  SampleUpdate,
  SpecStatus,
  ResultSource,
  TimelineEvent,
} from '../types/api'
import { http } from './client'

export interface SampleFilters {
  q?: string
  status?: SampleStatus[]
  priority?: SamplePriority
  product_id?: number
  client_id?: number
  responsible_id?: number
  received_from?: string
  received_to?: string
  page?: number
  size?: number
  sort?: string
}

export interface ResultFilters {
  spec_status?: SpecStatus
  source?: ResultSource
  test_code?: string
  sample_code?: string
  entered_from?: string
  entered_to?: string
  current_only?: boolean
  page?: number
  size?: number
  sort?: string
}

// Cada ação do workflow é uma rota própria (docs/api.md).
const ACTION_PATH: Record<SampleAction, string> = {
  start_analysis: 'start-analysis',
  submit_for_review: 'submit-for-review',
  approve: 'approve',
  reject: 'reject',
  return_to_analysis: 'return-to-analysis',
  cancel: 'cancel',
}

export type ActionPayload = { reason: string } | { password: string; comment?: string } | undefined

export const samplesApi = {
  search: (filters: SampleFilters) =>
    http.get<Page<SampleSummary>>('/samples', { ...filters }),
  get: (id: number) => http.get<SampleDetail>(`/samples/${id}`),
  create: (data: SampleCreate) => http.post<SampleDetail>('/samples', data),
  update: (id: number, data: SampleUpdate) => http.patch<SampleDetail>(`/samples/${id}`, data),
  assignTests: (id: number, testDefinitionIds: number[]) =>
    http.post<SampleDetail>(`/samples/${id}/tests`, { test_definition_ids: testDefinitionIds }),
  cancelTest: (sampleTestId: number, reason: string) =>
    http.post<SampleDetail>(`/sample-tests/${sampleTestId}/cancel`, { reason }),
  act: (id: number, action: SampleAction, payload?: ActionPayload) =>
    http.post<SampleDetail>(`/samples/${id}/${ACTION_PATH[action]}`, payload),
  timeline: (id: number, page = 1, size = 100) =>
    http.get<Page<TimelineEvent>>(`/samples/${id}/timeline`, { page, size }),
}

export const resultsApi = {
  enter: (sampleTestId: number, data: ResultEntry) =>
    http.post<ResultRead>(`/sample-tests/${sampleTestId}/results`, data),
  history: (sampleTestId: number) =>
    http.get<ResultRead[]>(`/sample-tests/${sampleTestId}/results`),
  search: (filters: ResultFilters) => http.get<Page<ResultListItem>>('/results', { ...filters }),
}
